# Licenses and References

FacsimiLab is built upon the incredible contributions of many open-source projects. We would like to recognize their efforts here. For packages that have a formal literature citation (i.e. in an academic journal or repository), we will try to cite the suggested reference. These citations will be added into the `references.bib`. Open source licenses will also be copied into this folder, as applicable.

## Licenses

- Nvidia CUDA (Docker Image): [NGC DEEP LEARNING CONTAINER LICENSE](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/LICENSE)
- Moby (Docker Engine): [Apache 2.0](https://github.com/moby/moby/blob/master/LICENSE)
- Micromamba (Docker Image): [Apache 2.0](https://github.com/mamba-org/micromamba-docker/blob/main/LICENSE)
- Quarto (Binary): [MIT License](https://quarto.org/license.html)
- Rclone (Docker Image): [MIT License](https://rclone.org/licence/)

## References

1. Allaire, J. J., Teague, C., Scheidegger, C., Xie, Y., & Dervieux, C. (2022). Quarto (1.2) [Computer software]. https://doi.org/10.5281/zenodo.5960048
2. Gayoso, A., Lopez, R., Xing, G., Boyeau, P., Valiollah Pour Amiri, V., Hong, J., Wu, K., Jayasuriya, M., Mehlman, E., Langevin, M., Liu, Y., Samaran, J., Misrachi, G., Nazaret, A., Clivio, O., Xu, C., Ashuach, T., Gabitto, M., Lotfollahi, M., … Yosef, N. (2022). A Python library for probabilistic analysis of single-cell omics data. Nature Biotechnology, 40(2), Article 2. https://doi.org/10.1038/s41587-021-01206-w
3. Lopez, R., Regier, J., Cole, M. B., Jordan, M. I., & Yosef, N. (2018). Deep Generative Modeling for Single-cell Transcriptomics. Nature Methods, 15(12), 1053–1058. https://doi.org/10.1038/s41592-018-0229-2
4. Team, T. P. D. (2023). pandas-dev/pandas: Pandas (v2.1.1) [Computer software]. Zenodo. https://doi.org/10.5281/ZENODO.3509134
5. Virshup, I., Bredikhin, D., Heumos, L., Palla, G., Sturm, G., Gayoso, A., Kats, I., Koutrouli, M., Scverse Community, Angerer, P., Bergen, V., Boyeau, P., Büttner, M., Eraslan, G., Fischer, D., Frank, M., Hong, J., Klein, M., Lange, M., … Theis, F. J. (2023). The scverse project provides a computational ecosystem for single-cell omics data analysis. Nature Biotechnology, 41(5), 604–606. https://doi.org/10.1038/s41587-023-01733-8
6. Waskom, M., Gelbart, M., Botvinnik, O., Ostblom, J., Hobson, P., Lukauskas, S., Gemperline, D. C., Augspurger, T., Halchenko, Y., Warmenhoven, J., Cole, J. B., Hoeven, E. T., De Ruiter, J., Vanderplas, J., Hoyer, S., Pye, C., Miles, A., Corban Swain, Meyer, K., … Williams, M. L. (2023). mwaskom/seaborn: V0.13.0 (September 2023) (v0.13.0) [Computer software]. Zenodo. https://doi.org/10.5281/ZENODO.592845
7. Wolf, F. A., Angerer, P., & Theis, F. J. (2018). SCANPY: Large-scale single-cell gene expression data analysis. Genome Biology, 19(1), 15. https://doi.org/10.1186/s13059-017-1382-0
